### PR TITLE
feature: Introduce `PhpUnitDataProviderReturnTypeFixer`

### PR DESCRIPTION
**Pull request by kubawerlos**

### CODE REVIEW

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (keradus) [inline doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: should we add to our ruleset?

is it that simply we (Fixer team) wants this change, or it's expected interface of provider method by PHPUnit ? (depends on this, different ruleset target)

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: Yes, I agreed with @Wirone that he will raise the PR after this one is merged (so we can focus here on the fixer here).

That way he will not have to see the diff there will be in data providers 😄

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: would be nice to see results of the fixer before merging it. maybe side PR that we apply changes of this fixer, but that side PR to be merged first? then, in this PR you can already enable it and we see it running in CI for our codebase

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: sure, here it is: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7072

File: tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php
Comment (Wirone) [inline tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php:242]: With positional placeholders in the template (`%1$s`) `$expected` could be passed only once.

**Review by Wirone — DISMISSED** [review]

File: tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php
Comment (kubawerlos) [reply tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php:81]: Consistency is king, right? `Foo\Bar` must be iterable.

Comment (Wirone) [reply tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php:66]: But isn't it static analyser's responsibility to ensure that defined return type actually _is_ `iterable`? It can be implicitly iterable: `class Bar implements IteratorAggregate` (explicitly iterable like `Foo\Bar&iterable` is [not possible](https://3v4l.org/U3XF8)). `Foo\Bar` may have generic annotations and provide information about iterated items, this would be lost. But since it's risky, maybe it's OK... 🤷‍♂️

TL;DR: what I mean is maybe this fixer should only replace `array` type or add `iterable` if no return type is defined. But then it's not risky anymore (probably) 😅.

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: Diff is smaller than I initially thought, because I expected that it will also replace arrays with yields 😆. I believe there was rule for that somewhere.. 🤔.

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: > should we add to our ruleset?
> 
> is it that simply we (Fixer team) wants this change, or it's expected interface of provider method by PHPUnit ? (depends on this, different ruleset target)

@kubawerlos ^^

Comment (SpacePossum) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: https://docs.phpunit.de/en/10.2/writing-tests-for-phpunit.html#data-providers

```
A data provider [...] must either return an array of arrays or an object that implements the Iterator interface.
```

so `dataprovider` may return `array` and not only `iterable` it is not a hard requirement by PHPUnit

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: I'd still stay with the claim that it is `iterable`: https://github.com/sebastianbergmann/phpunit/blob/10.2.2/src/Metadata/Api/DataProvider.php#L174-L194

File: tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php
Comment (kubawerlos) [reply tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php:66]: I wonder if this is the same level o "in theory" situation as we had the other day regarding `ConfigInterface`, the test case was added more to show how it behaves, not to ensure it behaves that way. Have you ever seen that a custom class was returned from the data provider?

What would you say we go with the current "everything to iterable" approach and if anyone reports "hey, may AwesomeDataProvider class type is changed to iterable" then we add "exclude" config option or something like that?

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: I believe @SpacePossum's comment was only about _expectation_ @keradus asked before. There isn't preferred way, PHPUnit just supports both. So if anything, it's not strict requirement, just convention to choose.

File: tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php
Comment (Wirone) [reply tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php:66]: Of course, it's an edge case (I have never seen something in my life 😉), we can do like you wrote.

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: Hard to tell if it is expected, I'd say it is a matter of personal taste, `iterable` is the contract. I can imagine a team of developers that agree to use `array` as much as one agreeing to use `\Generator`.

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: > https://docs.phpunit.de/en/10.2/writing-tests-for-phpunit.html#data-providers
> 
> ```
> A data provider [...] must either return an array of arrays or an object that implements the Iterator interface.
> ```

@SpacePossum It is outdated now 😉

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: > It must return [iterable](https://www.php.net/manual/en/language.types.iterable.php), either an array of arrays or an object that implements the Traversable interface.


hmm... let's quote it in Fixer description and put it in PHPUnit ruleset?

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: PHPUnit rulesets are for migrating to a higher PHPUnit version and this one is related to PHP, not PHPUnit version.

Don't we want it in `PhpCsFixer:risky`?

Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: @kubawerlos was `array|iterable` support in PHPUnit from day one or it's like that since specific version? Because maybe it's possible to define proper target set.

IMHO it could be added to `PhpCsFixer:risky`.

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: It was always that return type, PHPUnit 6 is the first one to support PHP 7+ only,.

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: but also PHPunit updated their docs to explicitly state `iterable` ? maybe we can take version for which they claim so?

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: ok, I'm also happy with PHPUnit 6, but maybe worth a comment in code about this decision.

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [inline src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:28]:
```suggestion
 * Fixer for data provider return time matching official expectation described in https://docs.phpunit.de/en/10.2/writing-tests-for-phpunit.html#data-providers : "It must return [iterable](https://www.php.net/manual/en/language.types.iterable.php), either an array of arrays or an object that implements the Traversable interface.".
 * Technically, this update was introduced to documentation of PHPUnit 10, yet it's technically possible already starting in PHPUnit 6.
 *
 * @author Kuba Werłos <werlos@gmail.com>
```

we do something similar for PSR rules, eg
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/src/Fixer/Whitespace/LineEndingFixer.php#L26-L31

Comment (keradus) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:28]: maybe it could partially be added also to `FixerDefinition` (non-blocking idea)

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: > It was always that return type, PHPUnit 6 is the first one to support PHP 7+ only,.


if someone would write code in PHP7 , but run PHPUnit 5 (which allow PHP7), it would still work for them, right?
imho as low as possible [so even before PHPUnit 5], or matching the version PHPUnit updated their docs - PHPUnit 10.

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: Let's make one step backward:

All the `@PHPUnitXXMigration` sets are to migrate from one PHPUnit version to another - higher - one, due to some classes changing name from underscores to namespaced, some assertions deprecated/removed/renamed/refactored or mocking changes.

This one is not related to PHPUnit **MIGRATION** more than (random fixers here) `PhpUnitInternalClassFixer` or `PhpUnitTestAnnotationFixer` - they are nice to have, but not related to any PHPUnit version.

_(So I've removed the fixer from `@PHPUnit60Migration:risky` set)_

Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: Maybe we need something like `@PHPUnitGoodPractices` 🤔?

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: I see a change on PHPUnit side.

PHPunit -9 was asking to have dataproviders returning array.
PHPUnit 10 started asking to have them returning iterable.

Even if code-level spec didn't changed, expectations named in docs did.

----

Maybe we can be lucky and get attention from @sebastianbergmann - Sebastian, if you noticed this question, what would be your take on it? Would you recommend to migrate code for dataproviders to return iterable explicitly?

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: It still would be "recommended", not required to migrate, like I guess using `function testFoo() {}` is recommended over `/** @test */ function foo() {}`.

As long as it is not required to migrate to a higher PHPUnit version I'd say let's not add it to migrations.

And definitely, this is something we can make in separate PR 😉

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (kubawerlos) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:28]: > Technically, this update was introduced to documentation

I'm not sure that looking at documentation over the actual implementation is a good DX. Even looking at the discussion here suggests we are not clear on what direction to follow.

Comment (kubawerlos) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:28]: Even in PHPUnit 4 all what code needed from data provider is to be able to iterate over it: https://github.com/sebastianbergmann/phpunit/blob/4.8.36/src/Util/Test.php#L374-L401

File: tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php
Comment (keradus) [inline tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php:217]: does it mean we are truncating this comment?
if so, i don't think we should - in other fixers, we were always taking extra look to preserve such comments

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (sebastianbergmann) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:2]: I am not aware of any change in PHPUnit 10 vs. PHPUnit 9 with regard to what a data provider is expected to return. PHPUnit just expects that whatever is returned can be iterated over and that each iteration yields an array.

File: tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php
Comment (kubawerlos) [reply tests/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixerTest.php]: Let's preserve the whole type in such case, unless `clean_namespace` is used, then it will be removed anyway.

File: doc/list.rst
Comment (keradus) [inline doc/list.rst:2474]: I believe we can now remove `testThatTestDataProvidersReturnIterableOrArray` method 🎉

Comment (kubawerlos) [reply doc/list.rst:2474]: Done

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [inline src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:60]: pls provide description with reasoning for using this fixer.

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [inline src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:35]:
```suggestion
            'The type returned by PHPUnit data provider must be `iterable`.',
```

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [inline src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:114]: I removed this block and unit tests for this fixer did not fail. What am I missing?

Comment (kubawerlos) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:114]: Hard to tell, are you using [VW Extension](https://github.com/hugues-m/phpunit-vw)? When I remove that block I'm getting failed tests: https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/commit/ded6e767abff695f56e4857f55b87523eee94516

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (keradus) [inline doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: please add a rationale why it should be this way / why fixer should be used

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [inline src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:113]:
```suggestion
                // continue;
```

test - is it only for 7.4?

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:114]: ack, it's only for 7.4

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [inline src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:112]:
```suggestion
            // @TODO: drop condition and it's body when PHP 8+ is required
            if ($tokens->generatePartialCode($typeStartIndex, $typeEndIndex) !== $typeAnalysis->getName()) {
```

Comment (keradus) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:112]: btw, @Wirone , this is example why you cannot downgrade tool to old PHP engine and claim it gonna work - even if it would execute, the tokenizer and syntax of file we want to fix is different under old PHP engine

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: I can't think of anything more than is in the summary. The is no general reason to use it other than to have unified code base - which I guess applies to 90% of the fixers.

TBH I disagree with the enforcement of the description added in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7128 (BTW I do not see any approval there 😉) - I cannot think for most of the fixers added there as exceptions about anything to have in the description other than what is the summary.

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: in `summary`, you say "it should return `iterable`". But JohnDoe, imagine a junior, will not have any clue why. In `description`, I want to have reasoning why/when this fixer should be used.

> BTW I do not see any approval there 😉

I believe I got some superpowers after 10 years in this repo 🤷🏻

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: > I want to have reasoning why/when this fixer should be used.

I don't have any :(

**Comment by kubawerlos** [comment]
There is no reason to have this fixer...

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (Wirone) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:113]: I don't get why - downgrading is only for replacing language features that are not supported on lower versions of the language. It does not change the logic or remove conditions. Both PHPstan and Rector are downgraded for release and works properly, even though some features are related to specific PHP versions that are higher than language version to which tool the is downgraded. Imagine the same codebase we have now, but using PHP8+ features, which after downgrade looks like... current code 🙂. Why wouldn't it work?

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: To be honest I also don't like merging without approvals. I understand @keradus is a project lead, but I believe there's always place for discussion about the vision 🙂.

Comment (kubawerlos) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: Especially with changes that are affecting the repo hard. Besides - as I already wrote - I disagree with the change, it increases the difficulty of adding a new fixer.

Like in [feature: AlphabeticalArrayKeySortFixer](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/5492), the [description](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/5492/files#diff-5dea671198fb463fcf94408fb01ac28d10b77ca9ae49ab0bb7e469b60619bb6fR57) is not what you, @keradus wrote: "I want to have reasoning why/when this fixer should be used."

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:113]: you can write code in PHP7.2, but if logic of that code is not covering PHP 7.2 syntax, then the code will not run properly.

So, imagine we say that we support only 8.2, there are no more dedicated test cases for PHP 7.x and extra conditions for specific 7.x syntax (like the one here) - even if you downgrade our source code, it wouldn't automatically add such extra condition

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: to trivialize the example - wouldn't it make sense to not only say "use === instead of ==", but also guide ppl *why* to do so?

**Comment by keradus** [comment]
> > I want to have reasoning why/when this fixer should be used.
> 
> I don't have any :(

> There is no reason to have this fixer...

There is, that's why it was created in first place. it's good to document it and that's all ;)

**Comment by kubawerlos** [comment]
> There is

And it's to have better (in a very subjective way) code for whoever might want to have it, like many other fixers.

The description goes in the documentation right below the summary - I don't understand why can't there be a single text parameter that would get a summary, very extended if needed.

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (Wirone) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:113]: But this has nothing to do with downgrading core code, but just technical decision which version is supported. If you decide to support 7.4, then you keep conditions in the code and test cases for 7.4, and downgrade to this version too. If you drop support for 7.4, then you remove conditions, tests and you change downgrading rules to go down to 8.0 only. So these are separate (but loosely related) concepts.

It works like this: you code in 8.2 and have such requirement, but during release you downgrade the code, change Composer requirements (so it can be used on lower PHP versions), test again if everything works and tag new version. Downgrading process has nothing to do with how fixers are covering PHP versions.

**Comment by keradus** [comment]
with one or 2 fields, most of our fixers are not describing why/when it's good to use them. it's not evangelising the good choices, it's only enforcing them

**Comment by keradus** [comment]
.

**Comment by Wirone** [comment]
In my opinion fixers' description should provide human-friendly description of what fixers are doing, because not everyone is happy with reviewing diffs from examples. If there's additional background for what fixer does, then it could be added, just like performance info or advantages of target syntax. But enforcing such description is not a good idea because many fixers are subjective, putting "because community decided like that" everywhere makes no sense 🤷‍♂️.

**Comment by keradus** [comment]
the PR was reverted 8mins ago.

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: How would you explain people why they should have ordered types? Or why they should have single empty line after declaring strict types? There are so many fixers that do things "just because" some group of people decided it should be a standard - what's the point in providing reasoning, if anyone can make their own reasoning out of the way how fixer works? Often it's just a matter of taste, not any particular reason.

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: often that taste is based on alignment - ie there were times when we literally read each line of PSR and made dedicated fixer for them.

eg https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7157

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (keradus) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:113]: > But this has nothing to do with downgrading core code, but just technical decision which version is supported.

for this project it was always the same decision. As with every bump of "which version of PHP we use to write our code", we were always getting rid of lot of legacy staff we no longer needed (eg conditions for old PHP tokenized content).

dependency is also interesting part - take our current dependencies and try to install them on 7.2 or 5.x. and I don't want to have a mess to support wide matrix of dependency versions.

Comment (keradus) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:113]: overall, I'm bringing to your visibility the complexity of maintaining code for low PHP tokenizer - it's not only how we transform our (sugar) syntax to lower version, but also that we would need to keep logic for those old tokenizers and for old dependency version.
This project is already complex enough and I do not want to have that extra complexity on top.

If you do believe it's worth to put effort to make newest PHP CS Fixer compatible and executable with legacy PHP version, do as you like to ;)
but I would recommend them to upgrade to new PHP engine instead of putting effort to use cutting-edge Fixer version

File: doc/rules/php_unit/php_unit_data_provider_return_type.rst
Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: Yeah, but fixer being part of PSR/PER ruleset is enough reasoning in my opinion, there's no need for clarify it in the description too.

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: and JohnDoe, having a look at `method_argument_space` only (eg because it was telling him that his codebase does not follow this rule, that is configured in project) has no clue about that.
Eg, `php-cs-fixer descbire` is not mentioning the ruleset, docs is claiming 4 different rulesets, and to understand the ultimate reason of this rule in first place (eg which community align on it first) - they need to go to codebase and start looking ,hoping for that.

ultimately, I could agree that it's non-obvious for some rules to say "why" other than "given community aligned on it". But we often don't say even that.
And for things like "=== vs ==", we do not tell any reason, only enforcing it, without supporting JohnDoe to understand it.

Imagine you ask junior dev in your team to follow a practice. whatever practice. if you enforce it on them from position of power, they will follow, but not be happy about it, and not liking it. But if you give them the reasoning (other than "do as I command"), they will understand it and see the value.

Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: I understand it, I just don't think enforcing such description is good, because it can lead to redundant and/or not helpful information 😉.

Comment (keradus) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: please provide any other way.

Comment (Wirone) [reply doc/rules/php_unit/php_unit_data_provider_return_type.rst:10]: For existing fixers:
1. Review existing fixers' descriptions
2. Create an issue with a todo list of fixers to improve
3. Provide PR(s) with improvements
4. Close the issue when every and each fixer is described properly

For new fixers: take care of it on code review stage.

File: src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php
Comment (Wirone) [reply src/Fixer/PhpUnit/PhpUnitDataProviderReturnTypeFixer.php:113]: @keradus you're mixing things and I believe we don't talk about the same. Since it's theoretical discussion, I'll just point basic things:

- I never meant to support 7.2 or even PHP5. What I was talking about is supporting _exact same_ versions like currently, but doing it while writing code that uses features from highest available version (for example native union types instead of phpDoc). To put it explicitly: write code on 8.2, downgrade it to 8.0 and support 8.0-8.2 (and starting to support 8.3). I don't know why you're talking about legacy PHP versions 🤷‍♂️.
- I did not bring downgrading thing up in terms of Fixer. I said it's good to have 2-way transformations because then it's possible to _use Fixer as a downgrade tool_. It's a topic for separate discussion.

**Pull request closed without merging**
